src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/JuelExpressionEvaluationProvider.java

```diff
@@ -16,27 +16,105 @@
 
 package org.powerflows.dmn.engine.evaluator.expression.provider;
 
+import de.odysseus.el.ExpressionFactoryImpl;
+import lombok.extern.slf4j.Slf4j;
 import org.powerflows.dmn.engine.evaluator.context.EvaluationContext;
+import org.powerflows.dmn.engine.evaluator.expression.provider.binding.ExpressionEvaluationException;
+import org.powerflows.dmn.engine.evaluator.expression.provider.juel.ContextFunctionMapper;
+import org.powerflows.dmn.engine.evaluator.expression.provider.juel.PowerflowsELContext;
 import org.powerflows.dmn.engine.model.decision.field.Input;
 import org.powerflows.dmn.engine.model.decision.rule.entry.InputEntry;
 import org.powerflows.dmn.engine.model.decision.rule.entry.OutputEntry;
 
+import javax.el.ELContext;
+import javax.el.ELException;
+import javax.el.ExpressionFactory;
+import javax.el.FunctionMapper;
 import java.io.Serializable;
 
+@Slf4j
 class JuelExpressionEvaluationProvider implements ExpressionEvaluationProvider {
 
+    private final ExpressionFactory expressionFactory;
+    private final FunctionMapper rootFunctionMapper;
+
+    JuelExpressionEvaluationProvider(final ExpressionEvaluationConfiguration configuration) {
+        this.expressionFactory = new ExpressionFactoryImpl();
+        rootFunctionMapper = new ContextFunctionMapper(configuration.getMethodBinding());
+    }
+
+    private PowerflowsELContext makeContext(final EvaluationContext evaluationContext) {
+        final PowerflowsELContext context = new PowerflowsELContext(rootFunctionMapper);
+        evaluationContext.getAll().forEach((key, value) ->
+            context.setVariable(key, expressionFactory.createValueExpression(value, value == null ? Object.class : value
+                    .getClass()))
+        );
+
+        return context;
+    }
+
+    private ELContext makeContext(final EvaluationContext evaluationContext, final InputEntry inputEntry) {
+        final PowerflowsELContext context = makeContext(evaluationContext);
+        context.setVariable(inputEntry.getNameAlias(), context.getVariableMapper()
+                .resolveVariable(inputEntry.getName()));
+
+        return context;
+    }
+
     @Override
     public Serializable evaluateInput(final Input input, final EvaluationContext evaluationContext) {
-        throw new UnsupportedOperationException();
+        log.debug("Starting evaluation of input: {} with evaluation context: {}", input, evaluationContext);
+
+        final Serializable result = evaluate((String) input.getExpression()
+                .getValue(), getResultType(input), makeContext(evaluationContext));
+
+        log.debug("Evaluated result: {}", result);
+
+        return result;
     }
 
+    private Class<? extends Serializable> getResultType(final Input input) {
+        if (input.getType() != null) {
```

Comment:
This condition is redundant, because Input builder keeps the validity by:

@Override
        protected Input assembleProduct() {
            if (product.expression == null) {
                product.expression = Expression.builder().build();
            }

            validateIsNonNull(product.name, "Name is required");
            validateIsNonNull(product.nameAlias, "Name alias is required");
            validateIsNonNull(product.type, "Type is required");

            return product;
        }

Reply:
right